migration/migrator/migrations/master/20210714135438_add_course_groups.py

```diff
@@ -0,0 +1,46 @@
+"""Migration for the Submitty master database."""
+from pathlib import Path
+from sqlalchemy import text
+import os
+
+def up(config, database):
+    """
+    Run up migration.
+
+    :param config: Object holding configuration details about Submitty
+    :type config: migrator.config.Config
+    :param database: Object for interacting with given database for environment
+    :type database: migrator.db.Database
+    """
+    database.execute("ALTER TABLE courses ADD COLUMN IF NOT EXISTS group_name varchar(255);")
+    database.execute("ALTER TABLE courses ADD COLUMN IF NOT EXISTS owner_name varchar(255);")
+    database.execute("ALTER TABLE courses DROP CONSTRAINT IF EXISTS group_validate;")
+    database.execute("ALTER TABLE courses DROP CONSTRAINT IF EXISTS owner_validate;")
+    database.execute("ALTER TABLE courses ADD CONSTRAINT group_validate CHECK (group_name ~ '^[a-zA-Z0-9_-]*$')")
+    database.execute("ALTER TABLE courses ADD CONSTRAINT owner_validate CHECK (owner_name ~ '^[a-zA-Z0-9_-]*$')")
+    semester_dir = Path(config.submitty['submitty_data_dir'], 'courses')
+    for semester in semester_dir.iterdir():
```

Comment:
This isn't safe to do on production.  We have directories from old terms that do not have corresponding Submitty databases / are not in the master database.

So we'll need to wrap this with some checks/exception handling.

Reply:
The UPDATE query is actually not dangerous or unsafe. If it does not find a course in the courses table then the query does nothing and moves on without any errors.